gnovm/pkg/integration/gno.go

```diff
@@ -0,0 +1,103 @@
+package integration
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	osm "github.com/gnolang/gno/tm2/pkg/os"
+	"github.com/rogpeppe/go-internal/testscript"
+)
+
+// SetupGno prepares the given testscript environment for tests that utilize the gno command.
+// If the `gno` binary doesn't exist, it's built using the `go build` command into the specified buildDir.
+// The function also include the `gno` command into `p.Cmds` to and wrap environment into p.Setup
+// to correctly set up the environment variables needed for the `gno` command.
+func SetupGno(p *testscript.Params, buildDir string) error {
+	// Try to fetch `GNOROOT` from the environment variables
+	gnoroot := os.Getenv("GNOROOT")
+	if gnoroot == "" {
+		// If `GNOROOT` isn't set, determine the root directory of github.com/gnolang/gno
+		goModPath, err := exec.Command("go", "env", "GOMOD").CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("unable to determine gno root directory")
+		}
+
+		gnoroot = filepath.Dir(string(goModPath))
+	}
+
+	if !osm.DirExists(buildDir) {
+		return fmt.Errorf("%q does not exist or is not a directory", buildDir)
+	}
+
+	// Determine the path to the gno binary within the build directory
+	gnoBin := filepath.Join(buildDir, "gno")
+	if _, err := os.Stat(gnoBin); errors.Is(err, os.ErrNotExist) {
+		// Build a fresh gno binary in a temp directory
+		gnoArgsBuilder := []string{"build", "-o", gnoBin}
+
+		// Forward `-covermode` settings if set
+		if coverMode := testing.CoverMode(); coverMode != "" {
+			gnoArgsBuilder = append(gnoArgsBuilder, "-covermode", coverMode)
+		}
+
+		// Append the path to the gno command source
+		gnoArgsBuilder = append(gnoArgsBuilder, filepath.Join(gnoroot, "gnovm", "cmd", "gno"))
+
+		if err = exec.Command("go", gnoArgsBuilder...).Run(); err != nil {
+			return fmt.Errorf("unable to build gno binary: %w", err)
+		}
+	} else if err != nil {
+		// Handle other potential errors from os.Stat
+		return err
+	}
```

Comment:
You can probably simplify this like:

```go
if _, err := os.Stat(gnoBin); err != nil {
	if !errors.Is(err, os.ErrNotExist) {
		// Handle other potential errors from os.Stat
		return err
	}

	// Build a fresh gno binary in a temp directory
	gnoArgsBuilder := []string{"build", "-o", gnoBin}

	// Forward `-covermode` settings if set
	if coverMode := testing.CoverMode(); coverMode != "" {
		gnoArgsBuilder = append(gnoArgsBuilder, "-covermode", coverMode)
	}

	// Append the path to the gno command source
	gnoArgsBuilder = append(gnoArgsBuilder, filepath.Join(gnoroot, "gnovm", "cmd", "gno"))

	if err = exec.Command("go", gnoArgsBuilder...).Run(); err != nil {
		return fmt.Errorf("unable to build gno binary: %w", err)
	}
}
```

Reply:
Resolved in:

https://github.com/gnolang/gno/pull/1249/commits/d3ec1fd32aaaed1703a228a6b4619bcb05e40ba1